platform/ios/src/MGLScaleBarView.m

```diff
@@ -0,0 +1,387 @@
+            label.font = [UIFont systemFontOfSize:8 weight:UIFontWeightMedium];
+            NSCharacterSet *characterSet = [[NSCharacterSet decimalDigitCharacterSet] invertedSet];
+            label.text = [[[self.formatter stringFromDistance:0] componentsSeparatedByCharactersInSet:characterSet] componentsJoinedByString:@""];
+            [label setNeedsDisplay];
+            [label sizeToFit];
+            [labels addObject:label];
+            [self addSubview:label];
+        }
+        _labels = labels;
+    }
+    return _labels;
+}
+
+#pragma mark - Layout
+
+- (void)layoutSubviews {
+    [super layoutSubviews];
+    
+    if (!self.row.lastObject || !self.row.lastObject.integerValue) {
+        // Current distance is not within allowed range
+        return;
+    }
+    
+    [self updateLabels];
+    [self layoutBars];
+    [self layoutLabels];
+}
+
+- (void)layoutBars {
+    CGFloat barWidth = (CGRectGetWidth(self.bounds) - self.borderWidth * 2.0f) / self.bars.count;
+    
+    NSUInteger i = 0;
+    for (UIView *bar in self.bars) {
+        CGFloat xPosition = barWidth * i + self.borderWidth;
+        bar.backgroundColor = (i % 2 == 0) ? self.primaryColor : self.secondaryColor;
+        bar.frame = CGRectMake(xPosition, self.borderWidth, barWidth, MGLBarHeight);
+        i++;
+    }
+    
+    self.containerView.frame = CGRectMake(CGRectGetMinX(self.bars.firstObject.frame),
+                                          CGRectGetMaxY(self.bounds)-MGLBarHeight,
+                                          self.actualWidth,
+                                          MGLBarHeight+self.borderWidth*2);
+    
+    self.containerView.layer.cornerRadius = 2.0f;
+    self.borderLayer.cornerRadius = 1.0f;
+    
+    [CATransaction begin];
+    [CATransaction setDisableActions:YES];
+    self.borderLayer.frame = CGRectInset(self.containerView.bounds, self.borderWidth, self.borderWidth);
+    self.borderLayer.zPosition = FLT_MAX;
+    [CATransaction commit];
+}
+
+- (void)layoutLabels {
+    CGFloat barWidth = self.bounds.size.width / self.bars.count;
+    NSUInteger i = 0;
+    for (MGLScaleBarLabel *label in self.labels) {
+        CGFloat xPosition = barWidth * i - CGRectGetMidX(label.bounds) + self.borderWidth;
+        label.frame = CGRectMake(xPosition, 0,
+                                 CGRectGetWidth(label.bounds),
+                                 CGRectGetHeight(label.bounds));
+        i++;
+    }
+}
+
+- (void)updateLabels {
+    NSArray *labels = [self.labels subarrayWithRange:NSMakeRange(1, self.labels.count-1)];
+    BOOL useMetric = [self usesMetricSystem];
+    NSUInteger i = 0;
+    
+    for (MGLScaleBarLabel *label in labels) {
+        CLLocationDistance barDistance = (self.row.firstObject.integerValue / self.row.lastObject.integerValue) * (i + 1);
+        if (!useMetric) { barDistance /= MGLFeetPerMeter; }
+        
+        label.text = [self.formatter stringFromDistance:barDistance];
```

Comment:
True, I guess we’d have a special case for when `-[NSLengthFormatter unitStringFromMeters:usedUnit:]` reports a different unit for the two labels. Avoiding repeated units can be tail work; I only suggested this approach as a space-saving measure, since the map view may not necessarily occupy the entire window.